cpp/daal/src/algorithms/engines/mrg32k3a/mrg32k3a_batch_impl.h

```diff
@@ -0,0 +1,114 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/*
+//++
+//  Implementation of the class defining the mrg32k3a engine.
+//--
+*/
+
+#include "algorithms/engines/mrg32k3a/mrg32k3a.h"
+#include "src/algorithms/engines/engine_batch_impl.h"
+#include "src/externals/service_rng.h"
+#include "src/data_management/service_numeric_table.h"
+
+namespace daal
+{
+namespace algorithms
+{
+namespace engines
+{
+namespace mrg32k3a
+{
+namespace internal
+{
+template <CpuType cpu, typename algorithmFPType = DAAL_ALGORITHM_FP_TYPE, Method method = defaultDense>
+class BatchImpl : public algorithms::engines::mrg32k3a::interface1::Batch<algorithmFPType, method>,
+                  public algorithms::engines::internal::BatchBaseImpl
+{
+public:
+    typedef algorithms::engines::mrg32k3a::interface1::Batch<algorithmFPType, method> super1;
+    typedef algorithms::engines::internal::BatchBaseImpl super2;
+    BatchImpl(size_t seed = 777) : baseRng(seed, __DAAL_BRNG_MRG32K3A), super2(seed) {}
+
+    void * getState() DAAL_C11_OVERRIDE { return baseRng.getState(); }
+
+    int getStateSize() const DAAL_C11_OVERRIDE { return baseRng.getStateSize(); }
+
+    services::Status saveStateImpl(byte * dest) const DAAL_C11_OVERRIDE
+    {
+        DAAL_CHECK(!baseRng.saveState((void *)dest), ErrorIncorrectErrorcodeFromGenerator);
+        return services::Status();
+    }
+
+    services::Status loadStateImpl(const byte * src) DAAL_C11_OVERRIDE
+    {
+        DAAL_CHECK(!baseRng.loadState((const void *)src), ErrorIncorrectErrorcodeFromGenerator);
+        return services::Status();
+    }
+
+    services::Status leapfrogImpl(size_t threadNum, size_t nThreads) DAAL_C11_OVERRIDE
+    {
+        int errcode = baseRng.leapfrog(threadNum, nThreads);
+        services::Status s;
+        if (errcode == __DAAL_LEAPFROG_METHOD_ERRCODE)
+            s.add(ErrorLeapfrogUnsupported);
+        else if (errcode)
+            s.add(ErrorIncorrectErrorcodeFromGenerator);
+        return s;
+    }
+
+    services::Status skipAheadImpl(size_t nSkip) DAAL_C11_OVERRIDE
+    {
+        int errcode = baseRng.skipAhead(nSkip);
+        services::Status s;
+        if (errcode == __DAAL_SKIP_AHEAD_METHOD_ERRCODE)
+            s.add(ErrorSkipAheadUnsupported);
+        else if (errcode)
+            s.add(ErrorIncorrectErrorcodeFromGenerator);
+        return s;
+    }
+
+    virtual BatchImpl<cpu, algorithmFPType, method> * cloneImpl() const DAAL_C11_OVERRIDE
+    {
+        return new BatchImpl<cpu, algorithmFPType, method>(*this);
+    }
+
+    bool hasSupport(engines::internal::ParallelizationTechnique technique) const DAAL_C11_OVERRIDE
+    {
+        switch (technique)
+        {
+        case engines::internal::family: return false;
+        case engines::internal::skipahead: return true;
+        case engines::internal::leapfrog: return true;
```

Comment:
Leap frog is not supported by MRG32K3A:
https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-vector-statistics-notes/2021-1/mrg32k3a.html